### ISSUE TITLE
Suggestion: Allow loss functions to accept per-label weights rather than per-sample

### ISSUE THREAD

**Issue by meder411** [pytorch/pytorch#3540]
I'd like to suggest allowing classification loss functions to accept a weight per-label as a parameter rather than a weight per-sample. Perhaps a dictionary of class labels as keys and weights as values would suffice. The big draw to PyTorch is how high-level it is, yet for networks that predict matrices (e.g. for segmentation), it is not trivial to use a weighted loss function. Maybe there is a better way (nothing in the docs or forums suggests it), but as far as I know the following example is the only way to do this.

If my output is a 8x1x500x500 labeling of 8 images with C labels, I have to create a 8x1x500x500 tensor where instead of a label I have that label's weight for every batch. Then, I must change the weights of the loss function at each batch, as
    
    for i in xrange(batches):
        ...
        my_new_weight_tensor = # Create the new weight tensor (possibly on the GPU)
        loss.weight = my_new_weight_tensor
        ...

In addition to being slightly cumbersome to code, this seems to use unnecessary memory and (possibly) communication overhead.

Instead I suggest the following user flow:

    w_dict = {0:0.25, 1:0.5, 2:1.} # Here C=3
    my_loss = BCELoss(weight=w_dict).cuda()

Now the weights are already on the GPU, and the calculation of the loss simply requires a lookup from the weights map.

I'm honestly surprised this isn't already how it's done. Although perhaps it is and I am just ignorant of it...

Also, even if there is no way to circumvent the memory usage (I suspect weighting with a tensor allows us to take advantage of fast matrix operations), the user flow would be substantially improved if the function constructs these tensors on the back-end.

**Comment by zou3519** [comment]
NLLLoss and CrossEntropyLoss accept per-class (per-label) weights. I'm not sure BCELoss should, the [docs](http://pytorch.org/docs/master/nn.html?highlight=bceloss#torch.nn.BCELoss) say that the target for BCELoss isn't a class label, it's a number between 0 and 1.

**Comment by meder411** [comment]
Thinking this over further, it seems that BCELoss should have a way to do this. Even through BCELoss with {0,1} targets basically reduces to NLLLoss, the annoying thing about the NLLLoss2d function is that I need a channel for each class. It'd be much more preferable to provide a Bx1xHxW binary image as my input to the loss.

**Comment by zou3519** [comment]
> BCELoss with {0,1} targets basically reduces to NLLLoss

Does it? Their formulation is different on the [docs](http://pytorch.org/docs/master/nn.html?highlight=bceloss#torch.nn.BCELoss).

>  the annoying thing about the NLLLoss2d function is that I need a channel for each class

The input to `NLLLoss2d` would have size `(N, 2, H, W)`, while the target should be `(N, H, W)`. Could you elaborate a little more on how this is annoying? I'm not sure I understand what you mean by providing a `Bx1xHxW` input.

**Comment by meder411** [comment]
After playing with this further, I suspect I wasn’t quite articulating my concern well at all and have resolved the issue. Feel free to close this.
